File: VB/MyTextEditWithCustomAutocomplete/CustomMaskManagerCommon.vb
Imports System
Imports DevExpress.Data.Mask

Namespace AutoCompleteMaskManager
	Public Class CustomMaskManagerCommon
		Inherits MaskManagerCommon(Of MaskManagerPlainTextState)

		Public Sub New(ByVal initialState As MaskManagerPlainTextState)
			MyBase.New(initialState)
		End Sub


		Protected Overrides Function CreateStateForApply(ByVal editText As String, ByVal cursorPosition As Integer, ByVal selectionAnchor As Integer) As MaskManagerPlainTextState
			Return New MaskManagerPlainTextState(editText, cursorPosition, selectionAnchor)
		End Function

		Protected Overrides Function GetEmptyState() As MaskManagerPlainTextState
			Return MaskManagerPlainTextState.Empty
		End Function

		Protected Overrides Function GetCursorPosition(ByVal state As MaskManagerPlainTextState) As Integer
			Return state.CursorPosition
		End Function

		Protected Overrides Function GetDisplayText(ByVal state As MaskManagerPlainTextState) As String
			Dim editText As String = GetEditText(state)
			Return editText
		End Function

		Protected Overrides Function GetEditText(ByVal state As MaskManagerPlainTextState) As String
			Return state.EditText
		End Function

		Protected Overrides Function GetSelectionAnchor(ByVal state As MaskManagerPlainTextState) As Integer
			Return state.SelectionAnchor
		End Function

		Public Overrides Function Backspace() As Boolean
			If IsSelection Then
				Return Insert(String.Empty)
			End If
			Dim head As String = GetCurrentEditText().Substring(0, DisplayCursorPosition)
			Dim tail As String = GetCurrentEditText().Substring(DisplayCursorPosition)
			If head.Length < 1 Then
				Return False
			End If
			Return Apply(New MaskManagerPlainTextState(head.Substring(0, head.Length - 1) & tail, DisplayCursorPosition - 1, DisplayCursorPosition - 1), StateChangeType.Delete)
		End Function

		Public Overrides Function Delete() As Boolean
			If IsSelection Then
				Return Insert(String.Empty)
			End If
			Dim head As String = GetCurrentEditText().Substring(0, DisplayCursorPosition)
			Dim tail As String = GetCurrentEditText().Substring(DisplayCursorPosition)
			If tail.Length < 1 Then
				Return False
			End If
			Return Apply(New MaskManagerPlainTextState(head & tail.Substring(1), DisplayCursorPosition, DisplayCursorPosition), StateChangeType.Delete)
		End Function

		Public Overrides Function Insert(ByVal insertion As String) As Boolean
			Dim changeType As StateChangeType = If(insertion.Length = 0 AndAlso IsSelection, StateChangeType.Delete, StateChangeType.Insert)

			Dim head As String = GetCurrentEditText().Substring(0, DisplaySelectionStart)

			Dim tail As String = GetCurrentEditText().Substring(DisplaySelectionEnd)

			Dim newText As String = head & insertion & tail


			Const SampleAutocompleteText As String = "_SampleAutocompleteText"


            If changeType = StateChangeType.Insert AndAlso tail.Length = 0 AndAlso insertion.Length = 1 Then
                Dim autocompletedText As String = newText & SampleAutocompleteText

                Return Apply(New MaskManagerPlainTextState(autocompletedText, newText.Length, autocompletedText.Length), changeType)
            Else
                Dim cursorPosition As Integer = (head & insertion).Length
                Return Apply(New MaskManagerPlainTextState(newText, cursorPosition, cursorPosition), changeType)
            End If
		End Function

		Public Overrides Sub SelectAll()
			CursorEnd(False)
			CursorHome(True)
		End Sub
	End Class
End Namespace

File: CS/MyTextEditWithCustomAutocomplete/CustomMaskManagerCommon.cs
using System;
using DevExpress.Data.Mask;

namespace AutoCompleteMaskManager {
    public class CustomMaskManagerCommon : MaskManagerCommon<MaskManagerPlainTextState> {
        public CustomMaskManagerCommon(MaskManagerPlainTextState initialState)
            : base(initialState) {
        }


        protected override MaskManagerPlainTextState CreateStateForApply(string editText, int cursorPosition, int selectionAnchor) {
            return new MaskManagerPlainTextState(editText, cursorPosition, selectionAnchor);
        }

        protected override MaskManagerPlainTextState GetEmptyState() {
            return MaskManagerPlainTextState.Empty;
        }

        protected override int GetCursorPosition(MaskManagerPlainTextState state) {
            return state.CursorPosition;
        }

        protected override string GetDisplayText(MaskManagerPlainTextState state) {
            string editText = GetEditText(state);
            return editText;
        }

        protected override string GetEditText(MaskManagerPlainTextState state) {
            return state.EditText;
        }

        protected override int GetSelectionAnchor(MaskManagerPlainTextState state) {
            return state.SelectionAnchor;
        }

        public override bool Backspace() {
            if (IsSelection) {
                return Insert(string.Empty);
            }
            string head = GetCurrentEditText().Substring(0, DisplayCursorPosition);
            string tail = GetCurrentEditText().Substring(DisplayCursorPosition);
            if (head.Length < 1) {
                return false;
            }
            return Apply(new MaskManagerPlainTextState(head.Substring(0, head.Length - 1) + tail, DisplayCursorPosition - 1, DisplayCursorPosition - 1), StateChangeType.Delete);
        }

        public override bool Delete() {
            if (IsSelection) {
                return Insert(string.Empty);
            }
            string head = GetCurrentEditText().Substring(0, DisplayCursorPosition);
            string tail = GetCurrentEditText().Substring(DisplayCursorPosition);
            if (tail.Length < 1) {
                return false;
            }
            return Apply(new MaskManagerPlainTextState(head + tail.Substring(1), DisplayCursorPosition, DisplayCursorPosition), StateChangeType.Delete);
        }

        public override bool Insert(string insertion) {
            StateChangeType changeType = (insertion.Length == 0 && IsSelection) ? StateChangeType.Delete : StateChangeType.Insert;

            string head = GetCurrentEditText().Substring(0, DisplaySelectionStart);

            string tail = GetCurrentEditText().Substring(DisplaySelectionEnd);

            string newText = head + insertion + tail;


            const string SampleAutocompleteText = "_SampleAutocompleteText";


            if (changeType == StateChangeType.Insert && tail.Length == 0 && insertion.Length == 1) {
                string autocompletedText = newText + SampleAutocompleteText;

                return Apply(new MaskManagerPlainTextState(autocompletedText, newText.Length, autocompletedText.Length), changeType);
            } else {
                int cursorPosition = (head + insertion).Length;
                return Apply(new MaskManagerPlainTextState(newText, cursorPosition, cursorPosition), changeType);
            }
        }

        public override void SelectAll() {
            CursorEnd(false);
            CursorHome(true);
        }
    }
}
